base/fastmath.jl

```diff
@@ -0,0 +1,326 @@
+         :atan => :atan_fast,
+         :atan2 => :atan2_fast,
+         :atanh => :atanh_fast,
+         :cbrt => :cbrt_fast,
+         :cis => :cos_fast,
+         :cos => :cos_fast,
+         :cosh => :cosh_fast,
+         :exp10 => :exp10_fast,
+         :exp2 => :exp2_fast,
+         :exp => :exp_fast,
+         :expm1 => :expm1_fast,
+         :hypot => :hypot_fast,
+         :lgamma => :lgamma_fast,
+         :log10 => :log10_fast,
+         :log1p => :log1p_fast,
+         :log2 => :log2_fast,
+         :log => :log_fast,
+         :max => :max_fast,
+         :min => :min_fast,
+         :minmax => :minmax_fast,
+         :sin => :sin_fast,
+         :sinh => :sinh_fast,
+         :sqrt => :sqrt_fast,
+         :tan => :tan_fast,
+         :tanh => :tanh_fast)
+
+const rewrite_op =
+    Dict(:+= => :+,
+         :-= => :-,
+         :*= => :*,
+         :/= => :/,
+         :^= => :^)
+
+function make_fastmath(expr::Expr)
+    op = get(rewrite_op, expr.head, :nothing)
+    if op != :nothing
+        var = expr.args[1]
+        rhs = expr.args[2]
+        if isa(var, Symbol)
+            # simple assignment
+            expr = :($var = $(op)($var, $rhs))
+        elseif isa(var, Expr) && var.head == :ref
+            # array reference
+            arr = var.args[1]
+            inds = tuple(var.args[2:end]...)
+            arrvar = gensym()
+            indvars = tuple([gensym() for i in inds]...)
+            expr = :($arrvar = $arr
+                     $indvars = $inds
+                     $(arrvar)[indvars] = $(op)($(arrvar)[$indvars], $rhs))
+        end
+    end
+    Expr(make_fastmath(expr.head), map(make_fastmath, expr.args)...)
+end
+function make_fastmath(symb::Symbol)
+    fast_symb = get(fast_op, symb, :nothing)
+    if fast_symb == :nothing
+        return symb
+    end
+    :(Base.FastMath.$(fast_symb))
+end
+make_fastmath(expr) = expr
+
+macro fastmath(expr)
+    make_fastmath(esc(expr))
+end
+
+
+
+# Basic arithmetic
+
+FloatTypes = Union(Float32, Float64)
+
+@fastmath for T in (Float32, Float64)
+    @eval begin
+        sub_fast(x::$T) = box($T,Base.neg_float_fast(unbox($T,x)))
+
+        add_fast(x::$T, y::$T) =
+            box($T,Base.add_float_fast(unbox($T,x), unbox($T,y)))
+        sub_fast(x::$T, y::$T) =
+            box($T,Base.sub_float_fast(unbox($T,x), unbox($T,y)))
+        mul_fast(x::$T, y::$T) =
+            box($T,Base.mul_float_fast(unbox($T,x), unbox($T,y)))
+        div_fast(x::$T, y::$T) =
+            box($T,Base.div_float_fast(unbox($T,x), unbox($T,y)))
+        rem_fast(x::$T, y::$T) =
+            box($T,Base.rem_float_fast(unbox($T,x), unbox($T,y)))
+
+        eq_fast(x::$T, y::$T) = Base.eq_float_fast(unbox($T,x),unbox($T,y))
+        ne_fast(x::$T, y::$T) = Base.ne_float_fast(unbox($T,x),unbox($T,y))
+        lt_fast(x::$T, y::$T) = Base.lt_float_fast(unbox($T,x),unbox($T,y))
+        le_fast(x::$T, y::$T) = Base.le_float_fast(unbox($T,x),unbox($T,y))
+    end
+end
```

Comment:
Could this be 

```julia
@fastmath begin

sub_fast{T<:FloatTypes}(x::T) = ....
...
end
````

Not that it will matter much, but `@eval` is one of the reasons building the standard library is slow.